userspace/plugin/plugin_loader.c

```diff
@@ -0,0 +1,238 @@
+	 * GetModuleHandle, however, returns the handle to a mapped module
+	 * without incrementing its reference count.
+	 *
+	 * This returns an HMODULE indeed, but they are the same thing
+	 */
+	return GetModuleHandle(path) != NULL;
+#else
+	/*
+	 * RTLD_NOLOAD (since glibc 2.2)
+	 *	Don't load the shared object. This can be used to test if
+	 *	the object is already resident (dlopen() returns NULL if
+	 *	it is not, or the object's handle if it is resident).
+	 *	This does not increment dlobject reference count.
+	 */
+	return dlopen(path, RTLD_LAZY | RTLD_NOLOAD) != NULL;
+#endif
+}
+
+bool plugin_check_required_api_version(const plugin_handle_t* h, char* err)
+{
+    uint32_t major, minor, patch;
+    const char *ver, *failmsg;
+    if (h->api.get_required_api_version == NULL)
+    {
+        strcpy(err, "plugin_get_required_api_version symbol not implemented");
+        return false;
+    }
+
+    ver = h->api.get_required_api_version();
+    if (sscanf(ver, "%" PRIu32 ".%" PRIu32 ".%" PRIu32, &major, &minor, &patch) != 3)
+    {
+        strcpy(err, "plugin provided an invalid required API version: '");
+        strncat(err, ver, PLUGIN_MAX_ERRLEN - 1);
+        strncat(err, "'", PLUGIN_MAX_ERRLEN - 1);
+        return false;
+    }
+
+    failmsg = NULL;
+    if(PLUGIN_API_VERSION_MAJOR != major)
+    {
+        failmsg = "': major versions disagree";
+    }
+    else if(PLUGIN_API_VERSION_MINOR < minor)
+    {
+        failmsg = "': framework's minor is less than the requested one";
+    }
+    else if(PLUGIN_API_VERSION_MINOR == minor && PLUGIN_API_VERSION_PATCH < patch)
+    {
+        failmsg = "': framework's patch is less than the requested one";
+    }
+
+    if (failmsg != NULL)
+    {
+        strcpy(err, "plugin required API version '");
+        strncat(err, ver, PLUGIN_MAX_ERRLEN - 1);
+        strncat(err, "' not compatible with the framework's API version '", PLUGIN_MAX_ERRLEN - 1);
+        strncat(err, PLUGIN_API_VERSION_STR, PLUGIN_MAX_ERRLEN - 1);
+        strncat(err, failmsg, PLUGIN_MAX_ERRLEN - 1);
+        return false;
+    }
+
+    return true;
+}
+
+plugin_caps_t plugin_get_capabilities(const plugin_handle_t* h)
+{
+    plugin_caps_t caps = CAP_NONE;
+
+    if (h->api.get_id != NULL
+        && h->api.get_event_source != NULL
+        && h->api.open != NULL
+        && h->api.close != NULL
+        && h->api.next_batch != NULL)
+    {
+        caps = (plugin_caps_t)((uint32_t) caps | (uint32_t) CAP_SOURCING);
+    }
+
+    if (h->api.get_fields != NULL
+        && h->api.extract_fields != NULL)
+    {
+        caps = (plugin_caps_t)((uint32_t) caps | (uint32_t) CAP_EXTRACTION);
+    }
+
+    return caps;
+}
+
+// little hack for simplifying the plugin_check_required_symbols function
+#define SYM_REQCHECK(a, e, s) \
+    if(a->api.s == NULL) \
+    { \
+        strcpy(e, "symbol not implemented: "); \
+        strncat(e, #s, PLUGIN_MAX_ERRLEN - 1); \
+        return false; \
+    }
```

Comment:
This is a nit, but you might want to wrap this in a dummy `do {} while(0)`, that way you make sure the whole macro doesn't interact erroneously with surrounding code blocks.
```suggestion
    do { \
        if(a->api.s == NULL) \
        { \
            strcpy(e, "symbol not implemented: "); \
            strncat(e, #s, PLUGIN_MAX_ERRLEN - 1); \
            return false; \
        } \
    } while(0)
```